ports/nrf/common-hal/alarm/time/TimeAlarm.c

```diff
@@ -0,0 +1,98 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+//#include "esp_sleep.h"
+
+#include "py/runtime.h"
+//#include "supervisor/esp_port.h"
+#include <stdio.h>
+
+#include "shared-bindings/alarm/time/TimeAlarm.h"
+#include "shared-bindings/time/__init__.h"
+
+void common_hal_alarm_time_timealarm_construct(alarm_time_timealarm_obj_t *self, mp_float_t monotonic_time) {
+    self->monotonic_time = monotonic_time;
+}
+
+mp_float_t common_hal_alarm_time_timealarm_get_monotonic_time(alarm_time_timealarm_obj_t *self) {
+    return self->monotonic_time;
+}
+
+mp_obj_t alarm_time_timealarm_get_wakeup_alarm(size_t n_alarms, const mp_obj_t *alarms) {
+    // First, check to see if we match
+    for (size_t i = 0; i < n_alarms; i++) {
+        if (MP_OBJ_IS_TYPE(alarms[i], &alarm_time_timealarm_type)) {
+            return alarms[i];
+        }
+    }
+    alarm_time_timealarm_obj_t *timer = m_new_obj(alarm_time_timealarm_obj_t);
+    timer->base.type = &alarm_time_timealarm_type;
+    // TODO: Set monotonic_time based on the RTC state.
+    timer->monotonic_time = 0.0f;
+    return timer;
+}
+
+bool alarm_time_timealarm_woke_us_up(void) {
+  return rtc_woke_up_counter;
+}
+
+void alarm_time_timealarm_reset(void) {
+    port_disable_interrupt_after_ticks_ch(1);
+    rtc_woke_up_counter = 0;
+}
+
+void alarm_time_timealarm_set_alarms(bool deep_sleep, size_t n_alarms, const mp_obj_t *alarms) {
+    bool timealarm_set = false;
+    alarm_time_timealarm_obj_t *timealarm = MP_OBJ_NULL;
+
+    for (size_t i = 0; i < n_alarms; i++) {
+        if (!MP_OBJ_IS_TYPE(alarms[i], &alarm_time_timealarm_type)) {
+            continue;
+        }
+        if (timealarm_set) {
+            mp_raise_ValueError(translate("Only one alarm.time alarm can be set."));
+        }
+        timealarm  = MP_OBJ_TO_PTR(alarms[i]);
+        timealarm_set = true;
+    }
+    if (!timealarm_set) {
+        return;
+    }
+
+    // Compute how long to actually sleep, considering the time now.
+    mp_float_t now_secs = uint64_to_float(common_hal_time_monotonic_ms()) / 1000.0f;
+    mp_float_t wakeup_in_secs = MAX(0.0f, timealarm->monotonic_time - now_secs);
+    int wsecs = (int)(wakeup_in_secs);
+    // timealarm is implemented by RTC, which is a 24bit counter
+    // running at 32768Hz. So, 2^24 / 32768 = 512sec is an upper limit.
+    if (wsecs >= 512) {
+        mp_raise_ValueError(translate("Alarm time must be < 512 seconds."));
+    }
+
+    uint32_t wakeup_in_ticks = (uint32_t)(wakeup_in_secs * 1024.0f);
+    port_interrupt_after_ticks_ch(1, wakeup_in_ticks);
```

Comment:
Instead of setting up the interrupt here, which is limited to 512 seconds, you can just remember how long you need to sleep. Then, actually do a `mp_hal_delay_ms()` or the equivalent in `common_hal_alarm_light_sleep_until_alarms()`.  More comments there.